database/rrdhost.c

```diff
@@ -664,6 +664,7 @@ void rrdhost_free(RRDHOST *host) {
     // free it
 
     freez((void *)host->tags);
+    free_host_labels(host->labels);
```

Comment:
The lock for the `host->labels` has not been acquired so this is unsafe.

Reply:
I'm not sure it is unsafe at this point. In theory the host data structure is unreachable here so that's why it can be freed as a whole as observed in the subsequent line:
`    freez(host);`

Reply:
Which is the point that it becomes unreachable - and how does it interact with the lock around the labels list?

I think we might have to change the sections that acquire the labels lock so that they acquire the host lock first.

Reply:
I had a more detailed look at the code and it is definitely a problem, there are different ways to fix it - please discuss which one would work best:
* Remove the separate lock for the labels completely and acquire the host lock to access the field.
* Acquire the host lock around every section that acquires the labels lock.
* Acquire the labels lock in `rdhost_free` before acquiring the host lock and then free it after the structure is unlinked.

Reply:
After a discussion with @mfundul we came to the conclusion that the problem is more general - there are a lot of places in Netdata, where host locks are not acquired but they should. I added host locks for host labels, but we need to check other possible places that miss host locks as well (it should be planned as a separate task).